### PR TITLE
Fix NonMonotonicSequenceException again (#9035)

### LINKED ISSUE #9035: 2.5.22294: MapRouteDrawer#drawCurvedPath:312 - org.apache.commons.math3.exception.NonMonotonicSequenceException
## Map
diplomacy

## Log Message
points 1 and 2 are not strictly increasing (13.44 >= 13.44)

## TripleA Version
2.5.22294

## Java Version
11.0.6

## Operating System
Windows 10

## Memory
Heap utilization statistics [MB]
Used Memory: 130
Free memory: 1449
Total memory: 1579
Max memory: 1579

## Stack Trace
```
Exception: org.apache.commons.math3.exception.NonMonotonicSequenceExceptionpoints 1 and 2 are not strictly increasing (13.44 >= 13.44)
java.lang.Exception
	at org.apache.commons.math3.util.MathArrays.checkOrder(MathArrays.java:521)
	at org.apache.commons.math3.util.MathArrays.checkOrder(MathArrays.java:538)
	at org.apache.commons.math3.util.MathArrays.checkOrder(MathArrays.java:549)
	at org.apache.commons.math3.analysis.interpolation.SplineInterpolator.interpolate(SplineInterpolator.java:81)
	at games.strategy.triplea.ui.panels.map.MapRouteDrawer.drawCurvedPath(MapRouteDrawer.java:312)
	at games.strategy.triplea.ui.panels.map.MapRouteDrawer.drawRoute(MapRouteDrawer.java:95)
	at games.strategy.triplea.ui.panels.map.MapPanel.paint(MapPanel.java:697)
	at java.desktop/javax.swing.JComponent.paintChildren(JComponent.java:907)
	at java.desktop/javax.swing.JComponent.paint(JComponent.java:1083)
	at java.desktop/javax.swing.JComponent.paintToOffscreen(JComponent.java:5255)
	at java.desktop/javax.swing.RepaintManager$PaintManager.paintDoubleBufferedImpl(RepaintManager.java:1643)
	at java.desktop/javax.swing.RepaintManager$PaintManager.paintDoubleBuffered(RepaintManager.java:1618)
	at java.desktop/javax.swing.RepaintManager$PaintManager.paint(RepaintManager.java:1556)
	at java.desktop/javax.swing.RepaintManager.paint(RepaintManager.java:1323)
	at java.desktop/javax.swing.JComponent._paintImmediately(JComponent.java:5203)
	at java.desktop/javax.swing.JComponent.paintImmediately(JComponent.java:5013)
	at java.desktop/javax.swing.RepaintManager$4.run(RepaintManager.java:865)
	at java.desktop/javax.swing.RepaintManager$4.run(RepaintManager.java:848)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:848)
	at java.desktop/javax.swing.RepaintManager.paintDirtyRegions(RepaintManager.java:823)
	at java.desktop/javax.swing.RepaintManager.prePaintDirtyRegions(RepaintManager.java:772)
	at java.desktop/javax.swing.RepaintManager$ProcessingRunnable.run(RepaintManager.java:1890)
	at java.desktop/java.awt.event.InvocationEvent.dispatch(InvocationEvent.java:313)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:770)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:721)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:715)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:85)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:740)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)


```

### PR DESCRIPTION
**Pull request by RoiEXLab**
This fixes #9035 for real this time.
Quite an interesting bug.

## Release Note

FIX|In rare edge cases a NonMonotonicSequenceException could occur when planning moves

### CODE REVIEW

File: game-app/game-core/src/main/java/games/strategy/triplea/ui/panels/map/MapRouteDrawer.java [hunk 1]
```diff
@@ -177,13 +177,10 @@ private void drawDirectPath(final Graphics2D graphics, final Point2D start, fina
    */
   protected double[] newParameterizedIndex(final Point2D[] points) {
     final double[] index = new double[points.length];
-    if (index.length > 0) {
-      index[0] = 0;
-    }
```
Comment (RoiEXLab) [inline game-app/game-core/src/main/java/games/strategy/triplea/ui/panels/map/MapRouteDrawer.java:182]: Looks like I was coding way too much C++ when I was originally writing this code.
In java obviously the array is initialized with 0.

File: game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java [hunk 2]
```diff
@@ -37,6 +37,13 @@
 import org.triplea.java.collections.IntegerMap;
 
 final class MapRouteDrawerTest {
+  /**
+   * This is a very special value because {@code 13.44 + Double.MIN_NORMAL == 13.44} evaluates to
+   * true. So we use the square of this value (because the code computes the square root) in order
```
Comment (DanVanAtta) [inline game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java:42]: Describing how/why this is a specific edge case would be useful. As a read I'm not sure if that commentary adds much value as I wonder about the edge case (what it is, what it effects). In combination with this, renaming 'Magic value' to something more descriptive would be a related change to generally be more descriptive.

Perhaps the comment could contrast the non-edge case value, in other words, why is `13.44 + Double.MIN_NORMAL` surprising, what would we expect it to equal?

Comment (RoiEXLab) [reply game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java:42]: Good point. Without the old code this comment does not make a lot of sense. I will try adding some context to this.

Regarding the variable name:
I must admit that the variable name was chosen mainly for 2 reasons.
1. I wasn't able to come up with a rather short and still descriptive name (that's why I added the javadoc in the first place)
2. MAGIC_VALUE is still short enough so we don't need a line break in the method source method below 😅 

Still not sure how to call it, if you have a good Idead, please let me know.

Comment (DanVanAtta) [reply game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java:42]: It's hard to suggest a value without knowing more about why it is a magic value. I think it's worth bearing in mind that the standards for a clean test is different from production code. A named constant could get out of hand if we did that for a half a dozen interesting values. I'd recommend to instead elide this problem, we only use the value on one line. Instead inline it and add a line or two of commentary above the `Arguments.of` line explaining why `13.44` is so particularly interesting.

Comment (RoiEXLab) [reply game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java:42]: Sounds like a good idea 👍🏼

**Comment by RoiEXLab** [comment]
@DanVanAtta I updated the description, let me know what you think

File: game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java [hunk 3]
```diff
@@ -137,7 +137,15 @@ void verifySequenceIsTrulyMonotonic(final Point2D[] points) {
             Arguments.of(new Double(1, 1)),
             Arguments.of(new Double(-1, -1)),
             Arguments.of(new Double(1, -1), new Double(1, -1), new Double(1, -1)),
-            Arguments.of(new Double(-1, -1), new Double(1, 1), new Double(0, 0)))
+            Arguments.of(new Double(-1, -1), new Double(1, 1), new Double(0, 0)),
+            // 13.44 is chosen specifically to test a rare edge case that used to cause an exception
```
Comment (DanVanAtta) [inline game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java:141]: The commentary is improved, but we still have a lot of it. This is quite a lot to read and understand & maintain. Perhaps we can still do better. Why is 13.44 all that special? Are there other special values? What is the range of valid values accepted by the computation function? What if we simply test nearly all of them with an increment of 0.01? Where would that land us with the runtime of the test case?

Comment (RoiEXLab) [reply game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java]: Ok, I did some experiments and it seems like 13.44 is not all that special because this bug occurs with almost every "large" positive integer.
So I changed the test case to use '1' als a number instead which tests the same thing and adjusted the comment accordingly.

File: game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java [hunk 4]
```diff
@@ -138,14 +138,10 @@ void verifySequenceIsTrulyMonotonic(final Point2D[] points) {
...
             Arguments.of(new Double(-1, -1), new Double(1, 1), new Double(0, 0)),
-            // 13.44 is chosen specifically to test a rare edge case that used to cause an exception
-            // in the past. We used to add a minimal positive double to indices to make them
-            // strictly increasing as required by commons-math. However in case of 13.44 the
-            // expression 13.44 + Double.MIN_NORMAL == 13.44 evaluates to true, so the sequence is
-            // no longer strictly increasing. So we use the square of this value (because the code
-            // computes the square root) in order to verify we don't get this issue again.
-            Arguments.of(
-                new Double(0, 0), new Double(0, 13.44 * 13.44), new Double(0, 13.44 * 13.44)))
+            // We used to add a minimal positive double to indices to make them strictly increasing
```
Comment (DanVanAtta) [inline game-app/game-core/src/test/java/games/strategy/triplea/ui/panels/map/MapRouteDrawerTest.java:141]: At this point, is the comment even really needed? It does not really matter what we 'used to do', someone reading this test case would only want to know why the test case is interesting and why we expect it to pass. Noting that the value was added for an 'edge' case I wonder if it's just unnecessary context at this point.
